mm/ktsan/tests_inst.c

```diff
@@ -155,8 +158,116 @@ static void stack_race_write(void *arg)
 
 static void kt_test_stack_race(void)
 {
-	kt_test(kt_nop, stack_race_write, stack_race_write,
-		true, "stack-race", true);
+	kt_test(kt_nop, kt_nop, stack_race_write, stack_race_write,
+		"stack-race", true, true);
+}
+
+/* ktsan test: racy-use-after-free */
+
+struct uaf_obj {
+	int data[32];
+};
+
+struct uaf_arg {
+	struct kmem_cache *cache;
+	struct uaf_obj *obj;
+};
+
+void kt_uaf_setup(void *p)
+{
+	struct uaf_arg *arg = (struct uaf_arg *)p;
+
+	arg->cache = kmem_cache_create("uaf_cache", sizeof(struct uaf_obj),
+					0, 0, NULL);
+	BUG_ON(!arg->cache);
+	arg->obj = kmem_cache_alloc(arg->cache, GFP_KERNEL);
+	BUG_ON(!arg->obj);
+}
+
+void kt_uaf_teardown(void *p)
+{
+	struct uaf_arg *arg = (struct uaf_arg *)p;
+
+	kmem_cache_destroy(arg->cache);
+}
+
+void kt_uaf_free(void *p)
+{
+	struct uaf_arg *arg = (struct uaf_arg *)p;
+
+	kmem_cache_free(arg->cache, arg->obj);
+}
+
+void kt_uaf_use(void *p)
+{
+	struct uaf_arg *arg = (struct uaf_arg *)p;
+
+	use(arg->obj->data[0]);
+}
+
+void kt_test_racy_use_after_free(void)
+{
+	kt_test(kt_uaf_setup, kt_uaf_teardown, kt_uaf_free, kt_uaf_use,
+		"racy-use-after-free", false, true);
+}
+
+/* ktsan test: SLAB_DESTROY_BY_RCU  */
+
+struct sdbr_obj {
+	int data[32];
+};
+
+struct sdbr_arg {
+	struct kmem_cache *cache;
+	struct sdbr_obj *obj;
+};
+
+static void sdbr_setup(void *p)
+{
+	struct sdbr_arg *arg = (struct sdbr_arg *)p;
+
+	arg->cache = kmem_cache_create("sdbr_cache", sizeof(struct sdbr_obj),
+					0, SLAB_DESTROY_BY_RCU, NULL);
+	BUG_ON(!arg->cache);
+	arg->obj = kmem_cache_alloc(arg->cache, GFP_KERNEL);
+	BUG_ON(!arg->obj);
+}
+
+static void sdbr_teardown(void *p)
+{
+	struct sdbr_arg *arg = (struct sdbr_arg *)p;
+
+	kmem_cache_destroy(arg->cache);
+}
+
+static void sdbr_obj_free(void *p)
+{
+	struct sdbr_arg *arg = (struct sdbr_arg *)p;
+	struct sdbr_obj *obj;
+
+	obj = rcu_dereference(arg->obj);
+	rcu_assign_pointer(arg->obj, NULL);
```

Comment:
Do nothing: done.

> All such slabs should have a ctor, we should clear the shadow before calling the ctor (do we?).
But if we're inside a rcu read critical section, and we got a reference to an object, which is being concurrently freed, we can still access the object after ctor is called in another thread, as I understand.